### ISSUE TITLE
Column offset and selection in `TableState`

### ISSUE THREAD

**Issue by achristmascarl** [ratatui/ratatui#1250]
## Problem
For tables with many columns, there isn't currently a way to offset the columns that are rendered or to select a column.

## Solution
Add `column_offset` and `column_selected` fields to `TableState` that are used when rendering, similar to how the [row] `offset` and `selected` are currently used.

**Comment by joshka** [comment]
There's a semi-related PR to this which has stalled out due to timing and some incompatible changes that were made in parallel: https://github.com/ratatui-org/ratatui/pull/516

It's a good idea generally that I'd like to see made to work. In particular, this would be useful to hook up to horizontal mouse scrolling or something equivalent (e.g. Shift + Arrow keys) in a demo.

**Comment by airblast-dev** [comment]
Can a maintainer assign this to me? 

I am thinking of making two separate PR's, one for applying styles for the selected column and one for allowing to set the highlight symbol next to the cell (similar to what the mentioned PR offered with `ColumnHighlightSpacing`) as I think doing the changes as seperate PR's would make it easier to review.

**Comment by orhun** [comment]
You got it!

**Comment by joshka** [comment]
I'm a big fan of smaller PRs that just do one thing well. They make it easy to review.

**Comment by airblast-dev** [comment]
I want to give a little update on my progress. Column selection, column highlighting, and cell highlighting are done. These are mostly ready for a PR.

## Problem:
I am not really sure how the column offset should behave with keeping the constraints (the widths field in `Table`) in mind. The point of allowing the user to set a column offset is to give the ability to navigate a large number of columns similar to how we can navigate many rows in `Table`, and `List`.

In the case of `Constraint::Max`, and `Constraint::Length` the intention of shrinking is explicit if the values are lesser than the columns minimally required width.  But with the other constraints (except `Constraint::Min`) we don't really now if shrinking the column is acceptable. We can check if the column could fit, but even then we can't be sure until the layout is created since the `Constraint` is an ambiguous width until the layout is processed. 

 I am not really sure what the correct behavior should be when an offset is set with constraints that contain the mentioned "ambiguous" constraints.

I think checking if the Column minimally fits (through a method such as `Cell::width`) and using its respective `Constraint`  if it does fit, seems the best way to do this. However even doing this isn't great as any externally computed and provided widths now have to be refactored (for example a user might want to allow every column to shrink, except the one that is selected instead of making it scrollable). This change would disallow shrinking were it is not explicit.

I think it is a bit confusing since constraints would work sort of differently than anywhere else in `ratatui`.

## Questions;
- Whats the preferred behavior to follow when computing the column widths keeping the constraints and offset in mind? 
- Should the offset be inline with how row offsets work?
- Should I create a PR for column selection, column highlighting, and cell highlighting? Considering that the offset situation is a bit complicated perhaps this is a good idea.

**Comment by joshka** [comment]
1. I'm not sure. Much of Ratatui is made in terms of a simple "Given this space, fit this stuff into it" rather than "Given this stuff, fit it into the following space". So the answer to how do i allow the table to expand beyond the visible space available is a difficult one that triggers some more fundamental questions.
2. I'm not sure what this question means
3. Assess whether there's likely to be overlap with the offset PR in ways that would lead to incompatible code changes. If there is then the PR should take that into account, but I suspect you can probably deliver column / cell highlighting distinct from the offset, so a PR is probably a good idea. When you do, can you please take a look at the existing PR and make notes about any previous discussion points that are addressed by your PR / changes to the way yours works compared to the existing PR (mostly to understand the intent of those changes).

**Comment by airblast-dev** [comment]
I could have been more clear in my second question. When rendering the rows, we get the start and end bound of what can and should be rendered. My idea was to do the same by rendering what can be fully rendered, but without any shrinking to the columns. Your answer to the first question also answers this, but I just wanted to clarify.

**Comment by matthewmturner** [comment]
Since the scrolling contents larger than viewable area is already done for rows does that mean that the issue for columns is because they are fundamentally different than rows or is it the combination of row and column logic that makes this complicated?

**Comment by joshka** [comment]
> Since the scrolling contents larger than viewable area is already done for rows does that mean that the issue for columns is because they are fundamentally different than rows or is it the combination of row and column logic that makes this complicated?

Yes, they are fundamentally different.

The table height is based on summing up the heights of all the rows.
If I have 200 rows that are all 1 high, then the total table height is 200.
To fit that into a 50 row screen you scroll to the offset and start rendering.

The table column widths are based on splitting up the known actual screen width using the layout constraints.
If I have 10 columns, each with a constraint of length(20), then the total width of the table is determined based entirely on the space that the table is rendered into. If the width is 100, then the column widths are compressed to each be 10. If the width is 500, then the widths are 20. This is just down to how the layout constraint system works. It's effectively "given a known space of X, chop it up with the provided constraints". You need more information to be able to render these constraints into a variable total width. A simple thought example: how wide are two columns in a table with the constraints: `[Min(1), Max(100)]`?
